[Breaking change]: .NET Linux applications do not look in `netcoredeps` sub-directory for native libraries (#45778)
dotnet/docs: docs/core/compatibility/interop/8.0/linux-netcoredeps.md

In earlier versions of .NET, Linux applications searched for native libraries in a `netcoredeps` subdirectory located next to the application executable. This behavior applied to all native library loads, including user-defined platform invokes (p/invokes).

## New behavior

In .NET 8 and later, Linux applications no longer search the `netcoredeps` subdirectory for native libraries. Native library resolution now follows standard mechanisms without relying on this subdirectory.

## Type of breaking change

This is a [source incompatible](../../categories.md#source-compatibility) change.

## Reason for change

The `netcoredeps` behavior was originally introduced to handle complex dependencies on third-party libraries in earlier .NET versions. Modern .NET versions no longer require this behavior due to improved dependency handling. Additionally, the mechanism isn't aligned with recommended practices for ELF platforms.

For more information, see [GitHub issue #114393](https://github.com/dotnet/runtime/issues/114393).

## Recommended action

If your application relied on the `netcoredeps` subdirectory for p/invokes or custom native library resolution, use the following alternatives:

> - Implement a custom resolution mechanism using <xref:System.Runtime.InteropServices.DllImportResolver> or the <xref:System.Runtime.Loader.AssemblyLoadContext.ResolvingUnmanagedDll> event.

Reviewer: There are two other (modern) options which can be used to make native loader look in directory of choice: `NativeLibrary.Load` and `NativeLibrary.SetDllImportResolver`.

> - Implement a custom resolution mechanism using <xref:System.Runtime.InteropServices.DllImportResolver> or the <xref:System.Runtime.Loader.AssemblyLoadContext.ResolvingUnmanagedDll> event.

Reviewer 2: ```suggestion
- Implement a custom resolution mechanism using the <xref:System.Runtime.Loader.AssemblyLoadContext.ResolvingUnmanagedDll> event or <xref:System.Runtime.InteropServices.NativeLibrary.SetDllImportResolver>.
```
- ResolvingUnmanagedDll is more likely to be solution that people are looking for, so I would mention it first
- DllImportResolver is just a delegate type that is not useful on its own. SetDllImportResolver docs explain what to do actually.

> - Implement a custom resolution mechanism using the <xref:System.Runtime.Loader.AssemblyLoadContext.ResolvingUnmanagedDll> event or <xref:System.Runtime.InteropServices.NativeLibrary.SetDllImportResolver>.

Reviewer: Looks like 42102332334721c91e94383c7fe6721fd15b7e6f was reverted

> - If an `RPATH` is required in your deployment, modify the ELF file explicitly.

Reviewer 2: ```suggestion
- If an `RPATH` is required in your deployment, modify the ELF file explicitly using `patchelf` utility.
```

## New behavior

In .NET 8 and later, Linux applications no longer search the `netcoredeps` subdirectory for native libraries. Native library resolution now follows standard mechanisms without relying on this subdirectory.

## Type of breaking change

This is a [source incompatible](../../categories.md#source-compatibility) change.

## Reason for change

The `netcoredeps` behavior was originally introduced to handle complex dependencies on third-party libraries in earlier .NET versions. .NET 8+ no longer requires this behavior due to improved dependency handling. Additionally, the mechanism isn't aligned with recommended practices for ELF platforms.

For more information, see [GitHub issue #114393](https://github.com/dotnet/runtime/issues/114393).

## Recommended action

If your application relied on the `netcoredeps` subdirectory for p/invokes or custom native library resolution, use the following alternatives:

- Implement a custom resolution mechanism using one of the following:

>     - <xref:System.Runtime.Loader.AssemblyLoadContext.ResolvingUnmanagedDll>

Reviewer 2: ```suggestion
    - <xref:System.Runtime.Loader.AssemblyLoadContext.ResolvingUnmanagedDll>
    - <xref:System.Runtime.Loader.AssemblyLoadContext.LoadUnmanagedDll>
```
We may want to mention LoadUnmanagedDll too

    - <xref:System.Runtime.InteropServices.NativeLibrary.SetDllImportResolver>
    - <xref:System.Runtime.InteropServices.NativeLibrary.Load>

>     - <xref:System.Runtime.InteropServices.NativeLibrary.SetDllImportResolver>

Reviewer 2: ```suggestion
```
This is duplicate.
